sdks/python/apache_beam/runners/portability/local_job_service.py

```diff
@@ -200,10 +202,13 @@ def run(self):
     if self._worker_id:
       env_dict['WORKER_ID'] = self._worker_id
 
-    p = subprocess.Popen(
-        self._worker_command_line,
-        shell=True,
-        env=env_dict)
+    # deadlock would happen with py2, so add a lock here.
+    # no problem with py3.
+    with SubprocessSdkWorker._lock:
+      p = subprocess.Popen(
```

Comment:
It turns out that deadlock happens with Popen(), rather than with p.wait().
Ran 1500 of wordcount.py with 3 workers and confirmed deadlock doesn't happen.